lib/friendly_shipping/services/tforce_freight/parse_create_bol_response.rb

```diff
@@ -16,26 +16,73 @@ class << self
+
+            rate_detail = json.dig("detail", "rateDetail")&.first
+            if rate_detail
+              service_code = rate_detail.dig("service", "code")
+              shipping_method = SHIPPING_METHODS.detect { |sm| sm.service_code == service_code }
+
+              rates = rate_detail["rate"].each_with_object({}) do |rate, result|
+                result[rate["code"].downcase.to_sym] = rate["value"].to_f
+              end
+
+              total_charges_value = rate_detail.dig("shipmentCharges", "total", "value")
+              total_charges_currency = rate_detail.dig("shipmentCharges", "total", "currency")
+              total_charges = Money.new(total_charges_value.to_f * 100, total_charges_currency || "USD")
+
+              billable_weight_value = rate_detail.dig("shipmentWeights", "billable", "value")
+              billable_weight_unit = rate_detail.dig("shipmentWeights", "billable", "unit")
+              billable_weight = Measured::Weight(billable_weight_value, billable_weight_unit&.downcase || :lb)
+
+              days_in_transit = rate_detail.dig("timeInTransit", "timeInTransit")&.to_i
+              cost_breakdown = build_cost_breakdown(rate_detail)
+            end
+
+            origin_service_center = json.dig("detail", "originServiceCenter")
+            email_sent = json.dig("detail", "pickup", "emailSent") == "true"
+            origin_is_rural = json.dig("detail", "pickup", "originIsRural") == "true"
+            destination_is_rural = json.dig("detail", "pickup", "destinationIsRural") == "true"
+
             documents = json.dig("detail", "documents", "image")&.map do |image_data|
               ParseShipmentDocument.call(image_data: image_data)
             end
 
             FriendlyShipping::ApiResult.new(
-              {
-                code: code,
-                message: message,
+              ShipmentInformation.new(
                 bol_id: bol_id,
                 pro_number: pro_number,
-                documents: documents
-              },
+                origin_service_center: origin_service_center,
+                email_sent: email_sent,
+                origin_is_rural: origin_is_rural,
+                destination_is_rural: destination_is_rural,
+                rates: rates,
+                total_charges: total_charges,
+                billable_weight: billable_weight,
+                days_in_transit: days_in_transit,
+                shipping_method: shipping_method,
+                documents: documents,
+                data: {
+                  cost_breakdown: cost_breakdown
+                }
+              ),
               original_request: request,
               original_response: response
             )
           end
+
+          private
+
+          def build_cost_breakdown(rate_detail)
+            {
+              "Rates" => rate_detail["rate"].each_with_object({}) do |rate, hash|
+                hash[rate["code"]] = rate["value"]
+              end,
```

Comment:
This is a nit but the spec confused me at first. Rates is a hash but the plural form makes it sound like it should be an array. TForce does the opposite and returns an array called "rate".

Reply:
Their response is all mixed up re expectations from the name of the key...

Array: detail, alerts, rate, commodities
Hash: summary, detailStatus, service, shipmentCharges, shipmentWeights, timeInTransit

Reply:
Yep. Fun, isn't it?